File: plugins/prompt-creator/scripts/validate_prompt.js
#!/usr/bin/env node
// validate_prompt.js — hearing JSON / trace JSON / generated prompt の最小検証 (Node 標準のみ)
"use strict";
const fs = require("fs");
const path = require("path");

function parseArgs(argv) {
  const args = { input: null, phase: null, schema: null };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--input") args.input = argv[++i];
    else if (argv[i] === "--phase") args.phase = argv[++i];
    else if (argv[i] === "--schema") args.schema = argv[++i];
  }
  return args;
}

function loadJson(p) {
  return JSON.parse(fs.readFileSync(p, "utf8"));
}

function defaultSchemaFor(phase) {
  const root = path.join(__dirname, "..");
  if (phase === "hearing") {
    return path.join(root, "skills", "run-prompt-elicit", "schemas", "hearing-result.schema.json");
  }
  if (phase === "sheet") {
    return path.join(root, "skills", "run-prompt-creator-7layer", "schemas", "hearing-result.schema.json");
  }
  if (phase === "trace") {
    return path.join(root, "skills", "run-prompt-create", "schemas", "build-trace.schema.json");
  }
  return null;
}

function checkRequired(obj, required, prefix) {
  const missing = [];
  for (const k of required) {
    if (obj[k] === undefined || obj[k] === null || obj[k] === "") {
      missing.push(`${prefix}.${k}`);
    }
  }
  return missing;
}

function detectPhase(input, explicitPhase) {
  if (explicitPhase) return explicitPhase;
  const ext = path.extname(input).toLowerCase();
  if (ext === ".json") return "trace";
  return "prompt";
}

function validateJsonBySchema(input, phase, schemaPath) {
  const data = loadJson(input);
  const sch = loadJson(schemaPath);
  let target = data;
  let required = sch.required || [];

  if (phase === "hearing") {
    target = data.phase1 || data;
    required = (sch.properties && sch.properties.phase1 && sch.properties.phase1.required)
      || sch.required || ["session_id", "timestamp", "answers"];
  }

  const missing = checkRequired(target, required, phase);
  if (missing.length > 0) {
    console.error("FAIL missing fields:");
    for (const m of missing) console.error(`  - ${m}`);
    process.exit(1);
  }
  console.log(`OK phase=${phase} schema=${schemaPath} required=${required.length} fields validated`);
}

function validatePromptText(input) {
  const text = fs.readFileSync(input, "utf8");
  const problems = [];
  for (let n = 1; n <= 7; n++) {
    const re = new RegExp(`#+\\s*Layer\\s*${n}\\s*[:：]`);
    if (!re.test(text)) problems.push(`Layer ${n}: marker missing`);
  }
  if (/\{\{[^}]+\}\}/.test(text)) problems.push("unexpanded placeholder remains");
  if (/TODO(?!\(human\))/i.test(text)) problems.push("TODO remains without TODO(human)");
  if (problems.length > 0) {
    console.error("FAIL prompt validation:");
    for (const p of problems) console.error(`  - ${p}`);
    process.exit(1);
  }
  console.log(`OK phase=prompt layers=7 file=${input}`);
}

function main() {
  const { input, phase, schema } = parseArgs(process.argv);
  if (!input) {
    console.error("usage: validate_prompt.js --input <file> [--phase hearing|sheet|trace|prompt] [--schema <path>]");
    process.exit(2);
  }
  const resolvedPhase = detectPhase(input, phase);
  if (resolvedPhase === "prompt") {
    validatePromptText(input);
    return;
  }
  const schemaPath = schema || defaultSchemaFor(resolvedPhase);
  if (!schemaPath || !fs.existsSync(schemaPath)) {
    console.error(`schema not found for phase=${resolvedPhase}: ${schemaPath || "(none)"}`);
    process.exit(2);
  }
  validateJsonBySchema(input, resolvedPhase, schemaPath);
}

main();

File: plugins/prompt-creator/skills/run-prompt-creator-7layer/scripts/validate-prompt.py
#!/usr/bin/env python3
# /// script
# name: validate-prompt
# purpose: hearing/sheet/trace JSON のスキーマ必須項目、または生成プロンプトの7層マーカー・未展開placeholder・TODO残存を検証する
# inputs:
#   - argv: --input <file> [--phase hearing|sheet|trace|prompt] [--schema <path>]
#   - file: --input の JSON または prompt テキスト、解決された schema JSON
# outputs:
#   - stdout: OK サマリ
#   - stderr: FAIL 詳細（不足フィールド / 検証問題）
#   - exit: 0=OK / 1=検証失敗 / 2=引数・schema 不在
# contexts: [C]
# network: false
# write-scope: none
# dependencies: []
# ///
# validate-prompt.py — hearing JSON / trace JSON / generated prompt の最小検証 (Python 標準のみ)
"""validate_prompt.js の python 移植。元の検証ロジック・終了コードを維持する。"""
import argparse
import json
import os
import re
import sys

# 旧 plugin-level scripts/ から skills/run-prompt-creator-7layer/scripts/ へ移動したため、
# プラグインルートは 3 階層上 (scripts/ → run-prompt-creator-7layer/ → skills/ → root)。
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))


def parse_args():
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument("--input")
    parser.add_argument("--phase")
    parser.add_argument("--schema")
    # A4-10: parse_known_args の黙殺を廃止 (failfast)。未知引数は argparse が exit 2。
    return parser.parse_args()


def load_json(p):
    with open(p, "r", encoding="utf-8") as f:
        return json.load(f)


def default_schema_for(phase):
    root = os.path.join(SCRIPT_DIR, "..", "..", "..")
    if phase == "hearing":
        return os.path.join(root, "skills", "run-prompt-elicit", "schemas", "hearing-result.schema.json")
    if phase == "sheet":
        return os.path.join(root, "skills", "run-prompt-creator-7layer", "schemas", "hearing-result.schema.json")
    if phase == "trace":
        return os.path.join(root, "skills", "run-prompt-create", "schemas", "build-trace.schema.json")
    return None


def check_required(obj, required, prefix):
    missing = []
    for k in required:
        v = obj.get(k) if isinstance(obj, dict) else None
        if v is None or v == "":
            missing.append(f"{prefix}.{k}")
    return missing


def detect_phase(input_path, explicit_phase):
    if explicit_phase:
        return explicit_phase
    ext = os.path.splitext(input_path)[1].lower()
    if ext == ".json":
        return "trace"
    return "prompt"


def validate_json_by_schema(input_path, phase, schema_path):
    data = load_json(input_path)
    sch = load_json(schema_path)
    target = data
    required = sch.get("required") or []

    if phase == "hearing":
        target = data.get("phase1") if isinstance(data, dict) and data.get("phase1") else data
        props = sch.get("properties") or {}
        phase1 = props.get("phase1") or {}
        required = phase1.get("required") or sch.get("required") or ["session_id", "timestamp", "answers"]

    missing = check_required(target, required, phase)
    if len(missing) > 0:
        sys.stderr.write("FAIL missing fields:\n")
        for m in missing:
            sys.stderr.write(f"  - {m}\n")
        sys.exit(1)
    print(f"OK phase={phase} schema={schema_path} required={len(required)} fields validated")


def validate_prompt_text(input_path):
    with open(input_path, "r", encoding="utf-8") as f:
        text = f.read()
    problems = []
    for n in range(1, 8):
        if re.search(rf"#+\s*Layer\s*{n}\s*[:：]", text) is None:
            problems.append(f"Layer {n}: marker missing")
    # 「出力指示」セクション以降は実行時入力プレースホルダー {{...}} を許可する
    # (seven-layer-markdown-template.md の「入力 placeholder は {{...}}」規定)。
    # それ以前の Layer 本文に残る {{...}} のみ未展開の骨格変数として検出する。
    layer_body = re.split(r"#+\s*出力指示", text, maxsplit=1)[0]
    if re.search(r"\{\{[^}]+\}\}", layer_body):
        problems.append("unexpanded placeholder remains")
    if re.search(r"TODO(?!\(human\))", text, re.IGNORECASE):
        problems.append("TODO remains without TODO(human)")
    if len(problems) > 0:
        sys.stderr.write("FAIL prompt validation:\n")
        for p in problems:
            sys.stderr.write(f"  - {p}\n")
        sys.exit(1)
    print(f"OK phase=prompt layers=7 file={input_path}")


def main():
    args = parse_args()
    input_path = args.input
    if not input_path:
        sys.stderr.write(
            "usage: validate-prompt.py --input <file> [--phase hearing|sheet|trace|prompt] [--schema <path>]\n"
        )
        sys.exit(2)
    resolved_phase = detect_phase(input_path, args.phase)
    if resolved_phase == "prompt":
        validate_prompt_text(input_path)
        return
    schema_path = args.schema or default_schema_for(resolved_phase)
    if not schema_path or not os.path.exists(schema_path):
        sys.stderr.write(
            f"schema not found for phase={resolved_phase}: {schema_path or '(none)'}\n"
        )
        sys.exit(2)
    validate_json_by_schema(input_path, resolved_phase, schema_path)


if __name__ == "__main__":
    main()
